Fusion RTOS porting round 3

kaleb-himes opened the pull request.

dgarske commented on IDE/VisualDSP/include.am:
```diff
@@ -0,0 +1,6 @@
+# included from Top Level Makefile.am
+# All paths should be given relative to the root
+
+EXTRA_DIST+= IDE/VisualDSP++/user_settings.h
```
These look wrong... `VisualDSP++`. First I think its `IDE/VisualDSP` and also we should avoid using '+' in file or directory names.

dgarske commented on IDE/VisualDSP/user_settings.h:
```diff
@@ -0,0 +1,752 @@
+    #define USE_WOLF_STRTOK
+    #define XSTRTOK(s1,d,ptr) wc_strtok((s1),(d),(ptr))
+
+    // TBD drowe: add a new FCL_STRNSTR and implement the new fclStrnstr code in common/clib/string/fclstrstr.c
```
Can we cleanup this comment? Not sure drowe should be in there. Also can we make sure and avoid `//` and use `/* */` syntax?

kaleb-himes replied:
Addressed

dgarske commented on src/internal.c:
```diff
@@ -77,8 +77,13 @@
     #endif
+        #ifdef FUSION_RTOS
+          #include <fclstdio.h>
+         #endif
```
Indent is off here

kaleb-himes replied:
Fixed

kaleb-himes replied:
Fixed

dgarske commented on src/sniffer.c:
```diff
@@ -25,17 +25,57 @@
+    #define XHTONL FNS_HTONL
+    #define XNTOHL FNS_NTOHL
+    #define XINADDR_NONE FNS_INADDR_NONE
+#else
```
If we are adding sniffer porting support please also add a user supplied macro option like `USE_CUSTOM_SNIFFX` that would allow the user to supply their own version of these.

kaleb-himes replied:
Added

kaleb-himes replied:
Added

dgarske commented on src/ssl.c:
```diff
@@ -29,6 +29,7 @@
     /* turn on GNU extensions for XVASPRINTF with wolfSSL_BIO_printf */
     #undef  _GNU_SOURCE
     #define _GNU_SOURCE
+    #include <stdarg.h>
```
Can we avoid adding this for all builds and put into the user_settings.h or wrap with some other macro specific to fusion or put in the settings.h case for the fusion rtos define?

kaleb-himes replied:
Moved to user_settings.h

dgarske commented on src/tls13.c:
```diff
@@ -1371,9 +1371,14 @@ int DeriveTls13Keys(WOLFSSL* ssl, int secret, int side, int store)
+#ifdef FUSION_RTOS
+        if (FCL_GETTIMEOFDAY(&now, 0) < 0)
+            return (word32)GETTIME_ERROR;
+            //TODO research why negative error code returns as word32
```
Can we resolve this TODO?

kaleb-himes replied:
Evaluated but I don't see a way to resolve, time wants to be non-negative so an unsigned return type is the only option for time calls across the spectrum perhaps we could consider returning simply WOLFSSL_FAILURE (IE zero) here instead of a negative return code that gets cast to an unsigned value.

kaleb-himes replied:
Attempted to return WOLFSSL_FAILURE but resulted in tests breaking and would likely effect user applications in use today. We'll need to revisit this in a stand-alone PR outside this porting effort. Did refactor this section however removing porting changes from the default and adding to a stand-alone section for the OS.

dgarske commented on src/wolfio.c:
```diff
@@ -694,7 +697,24 @@ int wolfIO_Recv(SOCKET_T sd, char *buf, int sz, int rdFlags)
+    recvd = (int)RECV_FUNCTION(sd, buf, sz, rdFlags, &err);
+
+    if(recvd < 0) {
+       if(err==63)
```
Can we avoid a hard coded `63`? Is there an error code lookup available or even just add one for this case? Please cleanup / resolve the TODO below at 714.

kaleb-himes replied:
Removed (customer ultimately used a custom I/O callback these changes were not necessary in the porting commit.

dgarske commented on src/wolfio.c:
```diff
@@ -703,8 +723,22 @@ int wolfIO_Recv(SOCKET_T sd, char *buf, int sz, int rdFlags)
+retry:
+    sent = (int)SEND_FUNCTION(sd, buf, sz, wrFlags, &err);
+    if(sent < 0) {
+       if(err==63)
```
Same here ... avoid hard coded not descriptive 63...

kaleb-himes replied:
Removed (customer ultimately used a custom I/O callback these changes were not necessary in the porting commit.

dgarske commented on src/wolfio.c:
```diff
@@ -694,7 +697,24 @@ int wolfIO_Recv(SOCKET_T sd, char *buf, int sz, int rdFlags)
 {
     int recvd;
 
+#ifdef FUSION_RTOS
```
Please add a custom `wolfIO_Recv` and `wolfIO_Send` for `FUSION_RTOS` and do not modify the general one...

kaleb-himes replied:
Removed (customer ultimately used a custom I/O callback these changes were not necessary in the porting commit.

dgarske commented on src/wolfio.c:
```diff
@@ -1462,8 +1498,14 @@ int EmbedOcspLookup(void* ctx, const char* url, int urlSz,
+            if (sfd != SOCKET_INVALID) {
+            #ifdef FUSION_RTOS
+                int err;
+                CloseSocket(sfd, &err);
```
If we don't have a macro for CloseSocket please add one. I think we do and this logic can be done in a single place via a macro.

kaleb-himes replied:
Addressed.

dgarske commented on wolfssl/wolfcrypt/wc_port.h:
```diff
@@ -487,6 +491,11 @@ WOLFSSL_API int wolfCrypt_Cleanup(void);
     #define XSEEK_END                VSEEK_END
     #define XBADFILE                 -1
     #define XFGETS(b,s,f)            -2 /* Not ported yet */
+    #define XVFPRINTF                vfprintf
```
Instead of adding all these variations just do an `#ifndef XVFPRINTF` at bottom for all these.

kaleb-himes replied:
Addressed

dgarske commented on wolfssl/wolfio.h:
```diff
@@ -372,6 +387,10 @@ WOLFSSL_API  int wolfIO_Recv(SOCKET_T sd, char *buf, int sz, int rdFlags);
     #define StartTCP()
+#elif defined(FUSION_RTOS)
+    #ifndef CloseSocket
+        #define CloseSocket FNS_CLOSE
```
Yes handle the additional error argument here in this macro....

kaleb-himes replied:
Done.

dgarske reviewed (changes requested):
This PR needs a lot of work

kaleb-himes commented:
Addressed all issues except the unsigned return time on a time value however did refactor that section to fit existing conventions instead of modifying the default.

kaleb-himes commented:
Addressed return value for time: Instead of returning a negative value cast as an unsigned value, will return WOLFSSL_FAILURE (IE zero) to indicate failure.

Addressed failure from New File detection test.

kaleb-himes commented:
Reverted the return value for time: causes tests to fail and may effect existing user applications. To be handled in a stand-alone PR and carefully with consideration for existing end-user applications.
